src/wfdirrd.c

```diff
@@ -1227,6 +1227,124 @@ DWORD DecodeReparsePoint(LPCWSTR szMyFile, LPCWSTR szChild, LPWSTR szDest, DWORD
 	return reparseTag;
 }
 
+#define PATH_PARSE_SWITCHOFF L"\\\\?\\" 
+#define PATH_PARSE_SWITCHOFF_SIZE (sizeof(PATH_PARSE_SWITCHOFF) - 1) / sizeof(wchar_t)
+#define REPARSE_MOUNTPOINT_HEADER_SIZE   8
+
+BOOL IsVeryLongPath(LPCWSTR pszPathName)
+{
+   return (wcslen(pszPathName) >= COUNTOF(PATH_PARSE_SWITCHOFF) - 1) && !wcsncmp(pszPathName, PATH_PARSE_SWITCHOFF, COUNTOF(PATH_PARSE_SWITCHOFF) - 1);
+}
+
+/* WFJunction
+ *
+ * Creates a NTFS Junction
+ * Returns either ERROR_SUCCESS or GetLastError()
+ */
+DWORD WFJunction(LPCWSTR pszLinkDirectory, LPCWSTR pszLinkTarget)
+{
+   DWORD		dwRet = ERROR_SUCCESS;
+   // Size assumption: We have to copy 2 path with each MAX_PATH long onto the structure. So we take 3 times MAX_PATH
+   char		reparseBuffer[MAX_PATH * 3];
+   WCHAR		szDirectoryName[MAX_PATH];
+   WCHAR		szTargetName[MAX_PATH];
+   PWCHAR	szFilePart;
+   DWORD		dwLength;
+
+
+   // Get the full path referenced by the target
+   if (!GetFullPathName(pszLinkTarget, MAX_PATH, szTargetName, &szFilePart))
+      return GetLastError();
+
+   // Get the full path referenced by the directory
+   if (!GetFullPathName(pszLinkDirectory, MAX_PATH, szDirectoryName, &szFilePart))
+      return GetLastError();
+
+   // Create the link - ignore errors since it might already exist
+   BOOL bDirCreated = CreateDirectory(pszLinkDirectory, NULL);
+   if (!bDirCreated) {
+      DWORD dwErr = GetLastError();
+      if (ERROR_ALREADY_EXISTS != dwErr)
+         return dwErr;
+      else {
+         // If a Junction already exists, we have to check if it points to the 
+         // same location, and if yes then return ERROR_ALREADY_EXISTS
+         wchar_t szDestination[MAX_PATH] = { 0 };
+         AppendToPath(pszLinkDirectory, szStarDotStar);
+         DecodeReparsePoint(pszLinkDirectory, NULL, szDestination, COUNTOF(szDestination));
+         StripFilespec(pszLinkDirectory);
+         if (!_wcsicmp(szDestination, pszLinkTarget)) {
+            SetLastError(ERROR_ALREADY_EXISTS);
+            return ERROR_ALREADY_EXISTS;
+         }
+      }
+   }
+
+   HANDLE hFile = CreateFile(
+      pszLinkDirectory,
+      GENERIC_WRITE,
+      0,
+      NULL,
+      OPEN_EXISTING,
+      FILE_FLAG_OPEN_REPARSE_POINT | FILE_FLAG_BACKUP_SEMANTICS,
+      NULL
+   );
+
+   if (INVALID_HANDLE_VALUE == hFile)
+      return GetLastError();
+
+   // Make the native target name
+   WCHAR szSubstituteName[MAX_PATH];
+
+   // The target might be
+   if (IsVeryLongPath(szTargetName)) {
+      // a very long path: \\?\x:\path\target
+      swprintf_s(szSubstituteName, MAX_PATH, L"\\??\\%s", &szTargetName[PATH_PARSE_SWITCHOFF_SIZE]);
+   } else {
+      if (szTargetName[0] == L'\\' && szTargetName[1] == L'\\')
+         // an UNC name: \\myShare\path\target
+         swprintf_s(szSubstituteName, MAX_PATH, L"\\??\\UNC\\%s", &szTargetName[2]);
+      else
+         // a normal full path: x:\path\target
+         swprintf_s(szSubstituteName, MAX_PATH, L"\\??\\%s", szTargetName);
+   }

```

Comment:
Just curious: this is saying that Winfile can internally use \\\\?\\ paths, but will never get here with \\??\\ paths or relative paths?

Reply:
Well it is trying to live with what comes around, and this was what I detected a during the last (LinkShellExtension) years.

I am rather sure about relative junctions. They don't exist, at least I was not able to create them. ( I played around a lot ...)